### PR TITLE
Stream OpenAI results

### PR DESCRIPTION
**Pull request by calyptobai**
This PR adds support for streaming OpenAI results for query explanations. This changes the API for `/answer` - the response is now streamed using HTTP SSE. For the first message, a JSON object is sent with the same data as before, but without the explanation text. Subsequent messages are incoming tokens, as streamed from OpenAI.

Response schema changes: (CC @anastasiya1155)

```diff
 struct AnswerResponse {
     user_id: String,
     query_id: uuid::Uuid,
     snippets: Vec<Snippet>,
-    selection: api::Response {
-       answer_path: String,
-       explanation: String,
-    }
+    answer_path: String,
  }
```

Note that we no longer return `explanation` as part of the `AnswerResponse`. The explanation field is streamed directly in the form of plaintext SSE messages which can be concatenated together to form the final explanation string.

### CODE REVIEW

File: server/bleep/src/webserver.rs [hunk 1]
```diff
@@ -128,19 +127,17 @@ where
     Json(Response::from(val))
 }
 
-pub(in crate::webserver) fn error<'a>(
-    kind: ErrorKind,
-    message: impl Into<Cow<'a, str>>,
-) -> Json<Response<'a>> {
+pub(in crate::webserver) type Result<T, E = Error> = std::result::Result<T, E>;
```
Comment (llogiq) [inline server/bleep/src/webserver.rs:130]: Did you know that `pub(in ..)` might be deprecated in the 2024 edition?

**Review by llogiq — APPROVED** [review]

File: server/bleep/src/webserver/answer.rs [hunk 2]
```diff
@@ -249,54 +244,58 @@ pub async fn handle(
...
     let answer_path = snippets.get(0).unwrap().relative_path.to_string();
 
-    Ok::<_, Json<super::Response<'static>>>(Json(super::Response::Answer(AnswerResponse {
-        snippets,
-        query_id,
-        user_id: params.user_id,
-        selection: api::Response {
-            answer: snippet_explanation,
+    let initial_event = Event::default()
+        .json_data(super::Response::Answer(AnswerResponse {
```
Comment (nerdypepper) [inline server/bleep/src/webserver/answer.rs:256]: can we implement `Into<super::Response> for AnswerResponse`, like we do for all these response types? like how `HoverableResponse` and `TokenInfoResponse` convert to `super::Response`. we could every macro-ify this impl in `webserver.rs`.

**Review by nerdypepper — APPROVED** [review]

File: server/bleep/src/webserver.rs (hunk 1 above)
Comment (calyptobai) [reply server/bleep/src/webserver.rs:130]: I didn't know that 😄 It's not particularly useful here, I'm just following the visibility specifier that other types use.

File: server/bleep/src/webserver/answer.rs [hunk 3]
```diff
@@ -452,11 +513,20 @@ Answer in GitHub Markdown:",
         prompt
     }
 
-    async fn select_snippet(&self, prompt: &str) -> Result<reqwest::Response, AnswerAPIError> {
-        self.send_until_success(prompt, 1, 0.0).await
+    async fn select_snippet(&self, prompt: &str) -> super::Result<String> {
```
Comment (ggordonhall) [inline server/bleep/src/webserver/answer.rs:516]: We could also maintain the non-streaming `answer-api` endpoint for these single token responses.

**Review by ggordonhall — APPROVED** [review]
Looks good, @anastasiya1155 are you adding the frontend changes to this branch?

**Review by llogiq — APPROVED** [review]

**Pull request merged**